File: abc/python/abc181/abc181_D.py
import sys
infile = sys.stdin.buffer

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    S = gs()
    ans = "No"
    if len(S) == 1 :
        if S[0] == "8" : ans = "Yes"
    else :
        c = [0] * 10
        for cc in S :
            x = int(cc)
            c[x] += 1
        if len(S) == 2 :
            for i in range(100) :
                if i % 8 > 0 : continue
                d1 = i % 10
                d2 = i // 10
                if d1 == d2 and c[d1] >= 2 : ans = "Yes"
                if d1 != d2 and c[d1] >= 1 and c[d2] >= 1 : ans = "Yes"
        else :
            for i in range(1000) :
                if i % 8 > 0 : continue
                d1 = i % 10; i2 = (i - d1) // 10
                d2 = i2 % 10; i3 = (i2 - d2) // 10
                d3 = i3 % 10
                if d1 == d2 and d1 == d3:
                    if c[d1] >= 3 : ans = "Yes"
                elif d1 == d2 :
                    if c[d1] >= 2 and c[d3] >= 1 : ans = "Yes"
                elif d1 == d3 :
                    if c[d1] >= 2 and c[d2] >= 1 : ans = "Yes"
                elif d2 == d3 :
                    if c[d2] >= 2 and c[d1] >= 1 : ans = "Yes"
                else :
                    if c[d1] >= 1 and c[d2] >= 1 and c[d3] >= 1 : ans = "Yes"
    sys.stdout.write(str(ans)+'\n')

if __name__ == '__main__' :
    main()
    sys.stdout.flush()

File: abc/go/abc181/abc181_D/abc181_D.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
)

type scanner struct { sc *bufio.Scanner }
func newScanner(input io.Reader) *scanner {
	sc := bufio.NewScanner(input)
	sc.Split(bufio.ScanWords)
	sc.Buffer(make([]byte, 1024), int(1e+9))
	return &scanner{sc}	
}
var rdr = newScanner(os.Stdin)
const BUFSIZE = 10000000
var wrtr = bufio.NewWriterSize(os.Stdout, BUFSIZE)
func gs() string  { rdr.sc.Scan(); return rdr.sc.Text() }

func main() {
	//f1, _ := os.Create("cpu.prof"); pprof.StartCPUProfile(f1); defer pprof.StopCPUProfile()
    defer wrtr.Flush()
	infn := ""
	if infn == "" && len(os.Args) > 1 {	infn = os.Args[1] }
    if infn != "" {	f, e := os.Open(infn); if e != nil { panic(e) }; rdr = newScanner(f) }
	S := gs(); ls := len(S); ans := "No"
	if ls == 1 {
		if S == "8" { ans = "Yes" }
	} else if ls == 2 {
		for _,s2 := range []string{"16","24","32","48","56","64","72","88","96"} {
			if S == s2 || S[1] == s2[0] && S[0] == s2[1] { ans = "Yes"}
		}
	} else {
		ccnt := make(map[byte]int)
		for _,c := range S { ccnt[byte(c)]++ }
		for i:=104; i<1000; i+=8 {
			sref := strconv.Itoa(i)
			s1,s2,s3 := sref[0],sref[1],sref[2]
			if s1 == '0' || s2 == '0' || s3 == '0' { continue }
			if ccnt[s1] < 1 || ccnt[s2] < 1 || ccnt[s3] < 1 { continue }
			if s1 == s2 && ccnt[s1] < 2 { continue }
			if s1 == s3 && ccnt[s1] < 2 { continue }
			if s2 == s3 && ccnt[s2] < 2 { continue }
			if s1 == s2 && s1 == s3 && ccnt[s1] < 3 { continue }
			ans = "Yes"
		}
	}
	fmt.Println(ans)
}



